src/storage/src/render/upsert.rs

```diff
@@ -765,52 +795,104 @@ where
                 input.next().await
             }
         } {
-            match event {
-                AsyncEvent::Data(cap, mut data) if output_cap.time().less_equal(cap.time()) => {
-                    stage_input(
-                        &mut stash,
-                        &mut data,
-                        &input_upper,
-                        &resume_upper,
-                        upsert_config.shrink_upsert_unused_buffers_by_ratio,
-                    );
-                }
-                AsyncEvent::Data(_cap, mut data) => {
-                    stage_input(
-                        &mut stash,
-                        &mut data,
-                        &input_upper,
-                        &resume_upper,
-                        upsert_config.shrink_upsert_unused_buffers_by_ratio,
-                    );
-                }
-                AsyncEvent::Progress(upper) => {
-                    // Ignore progress updates before the `resume_upper`, which is our initial
-                    // capability post-snapshotting.
-                    if PartialOrder::less_than(&upper, &resume_upper) {
-                        continue;
+            // Buffer as many events as possible. This should be bounded, as new data can't be
+            // produced in this worker until we yield to timely.
+            let events = [event]
+                .into_iter()
+                .chain(itertools::unfold(&mut input, |input| {
+                    input.next().now_or_never().flatten()
+                }))
```

Comment:
capturing works fine here
```suggestion
                .chain(std::iter::from_fn(|| input.next().now_or_never().flatten()))
```

Reply:
til